OpenTrons/opentrons-api
# feat(app, react-api-client): add "retry failed command" during Error Recovery

Closes [EXEC-432](https://opentrons.atlassian.net/browse/EXEC-432)



# Overview

Adds the ability to retry the command that caused the run to enter Error Recovery, resuming the run after the issued command completes. Implementing the retry logic requires laying out (hopefully) most of the remaining app-side foundational work, therefore that is included here as well.

There is a good bit of noise in this PR. A few specifics are mentioned via comments. Below are the major highlights and discussion points:

### Chaining Fixit Commands

In order to chain (or even create) commands with a `fixit` intent, we utilize `useRecoveryCommands`. I currently make the assumption that we do not want to handle fixit commands that fail during error recovery, so we always `continuePastCommandFailure`. Whether we actually want this behavior or not (or make this dependent on the exact fixit command) is an involved conversation and must also ensure error recovery can't possibly cause an infinite loop (such as a faulty sensor always detecting no tip attached after the user dispatches additional fixit pick up tip commands). 

### Identifying the Failed Run Command

Designs utilize splash pages that require knowledge of the failed command ([example](https://www.figma.com/file/8dMeu8MuPfXoORtOV6cACO/Feature%3A-Error-Recovery?type=design&node-id=324-157288&mode=design&t=rOQCfaqiD13cbARh-4)). While identifying and retaining the failed command is simple on the ODD, the complexity increases when we add desktop apps into the mix. A desktop user may be anywhere in the app and then may navigate to Error Recovery while a user engages (or has engaged) with Error Recovery on the ODD. In order to determine which command is the failed command, we must navigate backwards through the active run's list of commands, finding the most recent failed command with a `protocol` intent. And that's the issue; because `fixit` commands are added to the run command list, the following scenario may occur:

- User A on the ODD creates a significantly large number of `fixit` commands. These commands are appended to the protocol run commands list.
- User B on the desktop app navigates to the run, and the app attempts to find the last failed `protocol` command. No failed command is found when querying the server, which occurs due to the failed command being farther back in history than the `pageLength` sent to GET /runs/:runId/commands returned. [See this discussion](https://opentrons.slack.com/archives/C06M9F26YGH/p1715349271043819) for some background as to why this is the case.

@sanni-t's recent PR https://github.com/Opentrons/opentrons/pull/15031 solves the serialization issue for _completed_ runs. We could consider extending this functionality for commands during an on-going run (and maybe add a rate limiter?!?). @sfoster1 's thoughts are highlighted in the slack convo linked above after discussion in-person.

### Robot Motion State

If you look at [designs](https://www.figma.com/file/8dMeu8MuPfXoORtOV6cACO/Feature%3A-Error-Recovery?type=design&node-id=458-38190&mode=design&t=rOQCfaqiD13cbARh-4), there are a lot of different "robot in motion" screens. Sometimes we only want the motion screen when the robot does a command, and sometimes we want to persist it longer or indefinitely. I think the best way to handle motion is to explicitly control it by setting the motion screen before we do a command and then control it afterwards. This is a departure from the implicit `RobotInMotion` component takeover that all the other wizard flows use, but I don't think that's going to work for us, and even if it is possible, the control flow would be very hard to follow.



# Test Plan
- Turn on the BE/FE FFs for ER.
- Run any protocol with a Pick up tip command and don't pick up the tip. Proceed through the ER flows. Verify the initial Z-home occurs when launching ER, and clicking "resume" (see Review requests - designs are quite different than current flows and that's ok) causes a pick up tip command.
- NOTE: The retry failed command pick up tip will actually fail until https://github.com/Opentrons/opentrons/pull/15143 merges.
- Verify the run resumes. There may be some oddness with the command text on the desktop app/odd, but that's ok for now.


# Changelog
- Added "retry failed command" option in Error Recovery.


# Review requests
Designs have changed quite a bit in the past couple of weeks. This PR is just focused on functionality and structure!



# Risk assessment
low -- behind a ff



[EXEC-432]: https://opentrons.atlassian.net/browse/EXEC-432?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
